.github/bin/run-clang-tidy.sh

```diff
@@ -76,96 +111,110 @@ fi
+# Ideally loop until no changes in hashes, but one level good enough.
+
+# Create a sed expression that replaces header filenames with content hash.
+for f in $(cat ${FILES_OF_INTEREST}); do
+  if [ "${f##*.}" == "h" ]; then     # only headers expected to be relevant
+    c_hash="$(cat $f | md5sum | awk '{print $1}')"
+    echo "s|$f|$c_hash|g" | sed 's|\.|\\.|g' >> "${INCLUDE_REPLACE}"
+  fi
+done
 
-EXIT_CODE=0
-if [ -s ${FILES_TO_PROCESS} ]; then
-  echo "::group::Run ${PARALLEL_COUNT} parallel invocations of ${CLANG_TIDY} in chunks of ${FILES_PER_INVOCATION} files."
+# Create a content hash of each file, first changing include filenames with
+# their content-hash as described above.
+# Make a list of all files we don't have hashes for yet.
+for f in $(cat ${FILES_OF_INTEREST}); do
+  c_hash="$(sed -f "${INCLUDE_REPLACE}" < $f | md5sum | awk '{print $1}')"
+  echo "${c_hash}" >> "${HASHES_OF_INTEREST}"
+  if [ ! -r "${CONTENT_DIR}/${c_hash}" ]; then  # don't have, need to create
+    echo "${c_hash} ${f}" >> "${FILE_HASHES_TO_PROCESS}"
+  else
+    touch "${CONTENT_DIR}/${c_hash}"  # keep current to prevent garbage collect
+  fi
+done
 
-  cat ${FILES_TO_PROCESS} \
-    | xargs -P${PARALLEL_COUNT} -n ${FILES_PER_INVOCATION} -- \
-      ${CLANG_TIDY} --config="$(cat ${CLANG_CONFIG_FILE})" --quiet "${ADDITIONAL_TIDY_OPTIONS[@]}" 2>/dev/null \
-    | sed -e "s@${CANONICALIZE_SOURCE_PATH_REGEX}@@g" \
-    > ${TIDY_OUT}.tmp
+echo "$(wc -l < ${HASHES_OF_INTEREST}) files of interest."
 
-  mv ${TIDY_OUT}.tmp ${TIDY_OUT}
-  cat ${TIDY_OUT}
+if [ -s "${FILE_HASHES_TO_PROCESS}" ]; then
+  echo "::group::$(wc -l < ${FILE_HASHES_TO_PROCESS}) files to process"
+  cat "${FILE_HASHES_TO_PROCESS}"
   echo "::endgroup::"
 
-  echo ::group::Summary
-  sed -e 's|\(.*\)\(\[[a-zA-Z.-]*\]$\)|\2|p;d' < ${TIDY_OUT} | sort | uniq -c | sort -rn
+  # A compilation database is needed to run clang tidy.
+  echo ::group::Build compilation database
+  $(dirname $0)/make-compilation-db.sh
   echo "::endgroup::"
-else
-  echo "Skipping clang-tidy run: nothing to do"
+
+  # The files might be reported with various absolute prefixes that
+  # we remove to canonicalize and report as relative paths from project
+  # root.
+  readonly EXEC_ROOT="$(bazel info execution_root)"
+  readonly CURRENT_DIR="$(pwd)"
+  readonly CANONICALIZE_SOURCE_PATH_REGEX="^\(${EXEC_ROOT}\|${CURRENT_DIR}\|\.\)/"
+
+  # Helper script gets "<hash-value> <filename>" as single argument.
+  # Run clang-tidy on <filename> and store result in ${CONTENT_DIR}/<hash>
+  cat > "${CLANG_TIDY_HELPER}" <<EOF
+read OUTPUT_FILE INPUT_FILE <<< \$1
+${CLANG_TIDY} --quiet --config="$(cat ${CLANG_CONFIG_FILE})" \
+  ${ADDITIONAL_TIDY_OPTIONS[@]@Q} "\${INPUT_FILE}" 2>/dev/null \
+  | sed -e "s@${CANONICALIZE_SOURCE_PATH_REGEX}@@g" \
+  > "${CONTENT_DIR}/\${OUTPUT_FILE}.tmp"
+mv "${CONTENT_DIR}/\${OUTPUT_FILE}".tmp "${CONTENT_DIR}/\${OUTPUT_FILE}"
+EOF
+
```

Comment:
`chmod +x "${CLANG_TIDY_HELPER}"`
and call it directly?

Reply:
I prefer not having things executable if possible, and it would also mean that I have to add a `#!`.